client/clientutil/modelinfo.go

```diff
@@ -0,0 +1,461 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package clientutil
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/timeutil"
+)
+
+var (
+	// this list is a "nice" "human" "readable" "ordering" of headers to print
+	// off, sorted in lexical order with meta headers and primary key headers
+	// removed, and big nasty keys such as device-key-sha3-384 and
+	// device-key at the bottom
+	// it also contains both serial and model assertion headers, but we
+	// follow the same code path for both assertion types and some of the
+	// headers are shared between the two, so it still works out correctly
+	niceOrdering = [...]string{
+		"architecture",
+		"base",
+		"classic",
+		"display-name",
+		"gadget",
+		"kernel",
+		"revision",
+		"store",
+		"system-user-authority",
+		"timestamp",
+		"required-snaps", // for uc16 and uc18 models
+		"snaps",          // for uc20 models
+		"device-key-sha3-384",
+		"device-key",
+	}
+)
+
+type ModelFormatter interface {
+	LongPublisher(storeAccountID string) string
+	GetEscapedDash() string
+}
+
+// modelAssertJSON is a helper to write out a single assertion in json
+// format that matches how it's done in api_model.go. We have stolen
+// the structure from there.
```

Comment:
The structure must be in sync.

The thing is I'm a bit confused on where to put it - the structure is very simple and unlikely to change, and It feels overkill to create a new file for that, I did consider putting a comment above it in api_model.go that the structure should be in sync with the one in modelinfo.go.

I'll happily put it into a shared file, but I need some help on where it would make the best sense to put it

Reply:
It's not a call I should make, but it would seem to me that `asserts/model.go` might be the right place for it. The nice thing is that it's a package you are already including here, and it's also included in `daemon/api_model.go`.

If you also think that the suggestion is reasonable, you can make the change and then let's see what @pedronis says; or if in doubt, just keep it here and add a TODO comment (and also update the existing comment to make it clear that keeping the two structures in sync is essential).

Reply:
I think this is a very sensible suggestion, thanks for this, I have updated the code. Let's see if @pedronis agrees!